### PR TITLE
Add filter when checking out OpenSearch-Dashboards to speed up CI checks

### PR DESCRIPTION
**Pull request by cwperks**
### Description

Adds `filter` to the `actions/checkout@v2` step of `install-dashboards` to minimize the size of the repo that is cloned. The cypress folder is over 600MB in size and not needed for the security integ tests.

### Category

Test fix

### Issues Resolved

https://github.com/opensearch-project/security-dashboards-plugin/issues/1611

### Check List
- [ ] New functionality includes testing
- [ ] New functionality has been documented
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by cwperks** [comment]
The integ test failures are definitely due to high watermark.

```
Filesystem      Size  Used Avail Use% Mounted on
/dev/root        84G   74G  9.5G  89% /
```

Where is most of the space being consumed?

File: .github/workflows/integration-test.yml
Comment (cwperks) [inline .github/workflows/integration-test.yml:114]: I ran into the same issue described in this SO post: https://stackoverflow.com/a/70432217

Added `--runInBand --detectOpenHandles --forceExit` for debugging and to exit the tests. I'm not sure how else to resolve the test issue at this time.

```
Jest did not exit one second after the test run has completed.

This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
```

File: .github/workflows/integration-test.yml
Comment (cwperks) [inline .github/workflows/integration-test.yml:70]: There are many github actions on freeing up disk space. The description of this [action](https://github.com/marketplace/actions/free-disk-space-ubuntu) lists a few of them.

File: .github/workflows/integration-test.yml
Comment (DarshitChanpura) [reply .github/workflows/integration-test.yml:114]: aahh.. I ran into this too and I think these args resolved it on my end too. Seems like a good fix to me.

File: .github/workflows/integration-test.yml
Comment (DarshitChanpura) [inline .github/workflows/integration-test.yml:82]: what would be the behavior if there are more than these 4 or 1 of these doesn't exist?
Could you add a reference pointer using which you were able to list these 4?

**Review by DarshitChanpura — COMMENTED** [review]
thank you @cwperks !

File: .github/workflows/integration-test.yml
Comment (cwperks) [reply .github/workflows/integration-test.yml:82]: This is the full list I saw printed out from `choco list` on the runner

```
gradle 8.3.0
hg 5.0.0
imagemagick 7.1.1.1900
imagemagick.app 7.1.1.1900
InnoSetup 6.2.2
jq 1.7.0
julia 1.9.3
KB2919355 1.0.20160915
KB2919442 1.0.20160915
KB2999226 1.0.20181019
KB3033929 1.0.5
KB3035131 1.0.3
kubernetes-cli 1.28.2
kubernetes-helm 3.12.3
llvm 16.0.6
maven 3.8.7
mingw 12.2.0.3042023
Minikube 1.31.2
nginx 1.25.2
nodejs 18.18.0
nodejs.install 18.18.0
nssm 2.24.101.20180116
nuget.commandline 6.7.0
packer 1.9.4
php 8.2.11
pulumi 3.86.0
R.Project 4.3.1
rtools 4.3.5550
sbt 1.9.6
strawberryperl 5.32.1.1
swig 4.1.1
tortoisesvn 1.14.5.29465
vcredist140 14.36.32532
vswhere 3.1.7
wixtoolset 3.11.2
```

I haven't tested the case where one doesn't exist. I don't think it will fail the workflow, but I will check and reply back.

I didn't try to determine which of these were the largest, but I picked a handful that were pre-installed that I knew weren't required and chose to uninstall them.

**Review by peternied — CHANGES_REQUESTED** [review]
Thanks for digging into this - Could we merge just the change for filtering the checkout of dashboards, would that be enough to move the needle?  Alternatively can we adjust the highwatermark rules to be disabled which would make the fix simplier?

GitHub Action Marketplace [[search disk space]](https://github.com/marketplace?category=&type=actions&verification=&query=disk+space), couple of standouts:
- https://github.com/marketplace/actions/disk-space-optimizer Works on mac/windows?
- https://github.com/marketplace/actions/maximize-build-disk-space Seems very mature

File: .github/workflows/integration-test.yml
Comment (cwperks) [reply .github/workflows/integration-test.yml:99]: Low watermark is 90%. The tests don't use that much data, so I would recommend at least 1 GB (maybe less) below whatever 90% is on the runner.

File: .github/workflows/integration-test.yml
Comment (cwperks) [reply .github/workflows/integration-test.yml:76]: @DarshitChanpura let's look into other ways to clear up space or adjust the watermark. I don't think its possible to tell the github runner to not come pre-installed with bloatware (it may not necessarily be bloatware, but we don't need it for our CI) which would be my preferred fix :/

**Comment by cwperks** [comment]
@peternied 

> Could we merge just the change for filtering the checkout of dashboards, would that be enough to move the needle?

Unfortunately not :(. I tried that, but it still didn't clear enough space for both the Linux and windows runners.

> https://github.com/marketplace/actions/disk-space-optimizer Works on mac/windows?

I couldn't find any action that supported Windows well. Windows has not been easy to find documentation on.

File: .github/workflows/integration-test.yml
Comment (cwperks) [reply .github/workflows/integration-test.yml]: > what would be the behavior if there are more than these 4 or 1 of these doesn't exist?

It would still run if one of them didn't exist and not fail the workflow

**Comment by cwperks** [comment]
@peternied OSD Core disabled watermarking in this PR: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5172

i.e. `-E cluster.routing.allocation.disk.threshold_enabled=false`

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/.github/workflows/cypress_workflow.yml#L34

Even with watermarking disbled, they may still face an issue where the runner can run out of disk space in the future. I think it makes sense to address the space issue now instead of disabling the watermark, but unfortunately it looks like we will need to accept that the windows check will slow down fairly significantly.

**Comment by peternied** [comment]
@cwperks Can you get a run of these changes with 10 iterations that show its solid?  As much as I don't like the time penalty - I'd rather start from a known good place, and we can see about improving the 'cleanup' time seperately

File: .github/workflows/integration-test.yml
Comment (DarshitChanpura) [inline .github/workflows/integration-test.yml:17]: This change should be removed before final merge. Feel free to dismiss this review once the change is removed.

**Review by DarshitChanpura — CHANGES_REQUESTED** [review]

**Comment by cwperks** [comment]
@peternied Based on this comment (https://github.com/actions/runner-images/issues/1341#issuecomment-667478747) I think the long-term solution for windows may require changing the working directory to the C: drive from the default D: drive.

**Review by peternied — DISMISSED** [review]
@cwperks 10 iterations of passing jest tests is a great indicator to move forward.  Could you please fill an issue on the slow cleanup step for windows and we can revisit that in the future?

**Comment by cwperks** [comment]
4 / 10 of the windows runs failed on the first run, but succeeded on the 2nd. Its fairly stable with this fix, but still susceptible to failure. I'm filing an issue now around stability and speed for the CI on windows.

**Comment by cwperks** [comment]
@peternied issue on this repo: https://github.com/opensearch-project/security-dashboards-plugin/issues/1620

I am also going to file an issue in OSD core to see if its possible to only install prod dependencies.

Something like `NODE_ENVIRONMENT=production yarn osd bootstrap`

**Review by DarshitChanpura — DISMISSED** [review]

**Review by peternied — CHANGES_REQUESTED** [review]
> 4 / 10 of the windows runs failed on the first run, but succeeded on the 2nd. Its fairly stable with this fix, but still susceptible to failure. I'm filing an issue now around stability and speed for the CI on windows. 

Wait, what was the failure rate before?  I know you've been investing time in this - but I'd rather we make the minimal change with low downside rather that potential create more issues. 

Let me know if you've got before and after numbers - I am also happy to fold these changes into the overall CI improvements work and let you move onto other topics.

**Comment by cwperks** [comment]
@peternied The failure rate before this change was 100% on Windows.

**Comment by peternied** [comment]
On its own the change to filter check out of the Dashboards repository - happy to merge. 

Manually deleting/uninstall - I'm not comfortable without a 100% pass rate. It's hard to measure/maintain as a testament to the work you've done - I don't want us chasing future maintenance issues in the space if it isn't the full picture.

If we don't have the Windows platform operating reliably, it sounds like we should just disable it and then separately see how to get it reliable.

**Comment by cwperks** [comment]
> If we don't have the Windows platform operating reliably, it sounds like we should just disable it and then separately see how to get it reliable.

I would be supportive of that. This change would make the pass rate go from <20% to >70% so it would provide benefit in the interim while https://github.com/opensearch-project/security-dashboards-plugin/issues/1620 is worked on.

btw if any uninstall fails, the workflow continues as normal and would effectively work as it does today where it would run with all of the pre-installed software.

**Comment by DarshitChanpura** [comment]
@peternied @cwperks I've update the workflows to disable disk threshold check and have also reverted changes related to deleting pre-installed software on Windows runners

**Comment by cwperks** [comment]
Looks good to me!

**Review by peternied — APPROVED** [review]

**Comment by peternied** [comment]
@DarshitChanpura Thanks for your help on this.  I've created the code instability tracing issue [1] as well as a number of observations about test pass rate.  Lets try to keep working against a known state - while you are focused on this maybe you could update and track where we are at until we can get back to reliable.

- [1] https://github.com/opensearch-project/security-dashboards-plugin/issues/1624

**Review by DarshitChanpura — APPROVED** [review]

**Pull request merged**
